Implementing Ghost feature

Blenderiste09 opened the pull request.

Blenderiste09 committed:
Merge pull request #5 from NeKzor/auto-strafer

Syncing from NeKzor/auto-strafe

Blenderiste09 committed:
Merge branch 'master' of github.com:Blenderiste09/SourceAutoRecord

Blenderiste09 committed:
Merge pull request #6 from NeKzor/auto-strafer

Fix potential crashes with Server::GetPlayer

Blenderiste09 committed:
Fix missing files

Blenderiste09 committed:
Merge branch 'master' of github.com:NeKzor/SourceAutoRecord

Blenderiste09 committed:
Implemented fake strafing

Blenderiste09 committed:
Merge branch 'master' of github.com:NeKzor/SourceAutoRecord

Blenderiste09 committed:
Update Cheats.cpp

Blenderiste09 committed:
Including turning with velocity gain in description

It should be a bit more clear

Blenderiste09 committed:
sar_tas_setang

Blenderiste09 committed:
Fix debug ouput spamming while Tasing

Blenderiste09 committed:
Implemented crouch strafing

You can now strafe while being crouched.
Maximum speed is 81.314 ups

Blenderiste09 committed:
Added interpolation

sar_tas_setang and sar_tas_aim_at_point have a new optional parameter [speed].
Using it will make an interpolation.

Blenderiste09 committed:
Displaying each velocity component

sar_hud_velocity 3

Blenderiste09 committed:
Merge pull request #7 from Blenderiste09/Setang

Merge Setang into master

Blenderiste09 committed:
Merge branch 'master' of github.com:NeKzor/SourceAutoRecord

Blenderiste09 committed:
Revert "Merge branch 'master' of github.com:NeKzor/SourceAutoRecord"

This reverts commit a296642579cd73883a89d7949ded1985870db168, reversing
changes made to e1a1aab3e8f8eb5b3e4ec692af4f6bf8659a602f.

Blenderiste09 committed:
Re-Merging "master" branch from NeKzor

Because I'm dumb

Blenderiste09 committed:
Implenting speed gel strafing

Use sar_tas_strafe_vectorial 0 when strafing on speed gel

Blenderiste09 committed:
Regex forr Tasing

Blenderiste09 committed:
Merge branch 'master' of github.com:NeKzor/SourceAutoRecord

Blenderiste09 committed:
Fix include missing

Blenderiste09 committed:
Fixed ui_loadingscreen_transition_time

Blenderiste09 committed:
Merge branch 'master' of github.com:Blenderiste09/SourceAutoRecord into pr/13

Blenderiste09 committed:
Removed sar_tas_setang_rand

Blenderiste09 committed:
Merge branch 'master' of github.com:NeKzor/SourceAutoRecord

Blenderiste09 committed:
Implemented Ghost feature

Plays ghost of the selected demo.
sar_ghost_set_demo <demo file>
sar_ghost_set_CM_time <xx.yy>
sar_ghost_autostart <1|0>

Blenderiste09 committed:
Crash fix

Blenderiste09 committed:
Fixed copy.bat

Blenderiste09 committed:
Last fix copy.bat lol

Blenderiste09 committed:
Code cleaning

Blenderiste09 committed:
Crash fix

Position update function was still working even after a load even though the ghost didn't spawn yet

Blenderiste09 committed:
Added some customization

sar_ghost_height
sar_ghost_transparency

Also added the rotation update

Blenderiste09 committed:
Space fix

Blenderiste09 committed:
Start working on coop support

NeKzor commented on copy.bat:
```diff
@@ -1,18 +1,5 @@
-REM copy /Y "%src%" "%dest%\The Stanley Parable\%binary%"
-REM copy /Y "%src%" "%dest%\The Beginners Guide\%binary%"
-REM copy /Y "%src%" "%dest%\INFRA\%binary%"
+SET "dest=E:\Steam\steamapps\common"
```
Please don't commit this file when you edit it.

NeKzor commented on src/Features/Tas/CommandQueuer.hpp:
```diff
@@ -2,6 +2,7 @@
 #include <regex>
 #include <string>
 #include <vector>
+#include <regex>
```
Ehem.

NeKzor commented on src/Games/Linux/Portal2.cpp:
```diff
@@ -116,6 +116,11 @@ void Portal2::LoadOffsets()
     GetIServerEntity = 2; // CServerTools
     m_EntPtrArray = 48; // CServerTools::GetIServerEntity
+    SetKeyValueChar = 12; // CServerTools::SetKeyValue (const char *szValue)
+    SetKeyValueChar = 13; // CServerTools::SetKeyValue (float flValue )
```
SetKeyValueFloat?

NeKzor commented on src/Games/Linux/Portal2.cpp:
```diff
@@ -116,6 +116,11 @@ void Portal2::LoadOffsets()
     m_EntPtrArray = 48; // CServerTools::GetIServerEntity
+    SetKeyValueChar = 12; // CServerTools::SetKeyValue (const char *szValue)
+    SetKeyValueChar = 13; // CServerTools::SetKeyValue (float flValue )
+    SetKeyValueChar = 14; // CServerTools::SetKeyValue (const Vector &vecValue )
```
SetKeyValueVector?

NeKzor commented on src/Modules/Server.cpp:
```diff
@@ -269,6 +270,36 @@ DETOUR_STD(void, Server::GameFrame, bool simulating)
+        if (ghost->isPlaying) {
+            QAngle position = ghost->positionList.at(server->tickCount);
+            std::string position_string = std::to_string(position.x) + " " + std::to_string(position.y) + " " + std::to_string(position.z + sar_ghost_height.GetFloat());
+            server->SetKeyValueChar(ghost->ghost_entity, "origin", position_string.c_str());
```
Instead of concatenating strings why not use SetKeyValueVector function here?

NeKzor commented on src/Modules/Server.cpp:
```diff
@@ -269,6 +270,36 @@ DETOUR_STD(void, Server::GameFrame, bool simulating)
+
+			QAngle angles = ghost->angleList.at(server->tickCount);
+            std::string angle_string = std::to_string(angles.x) + " " + std::to_string(angles.y) + " " + std::to_string(angles.z);
+            server->SetKeyValueChar(ghost->ghost_entity, "angles", angle_string.c_str());
```
Same as above.

NeKzor commented on src/Modules/Server.cpp:
```diff
@@ -269,6 +270,36 @@ DETOUR_STD(void, Server::GameFrame, bool simulating)
+        }
+		if (server->tickCount == ghost->positionList.size()) {
+            console->Print("Ghost has finished.\n");
+            ghost->Reset();
```
All of this should be in a "Think" function of Ghost class.

NeKzor commented on src/Modules/Server.hpp:
```diff
@@ -21,14 +21,28 @@ class Server : public Module {
 
+    bool mapSpawning;
+    bool inMap;
+    int tickCount;
```
This should be moved to Ghost class.

Blenderiste09 committed:
Requested changes

Blenderiste09 committed:
Added sar_ghost_time_offset

Blenderiste09 committed:
Add name to ghost entity

In case we need it later (like for trail or something else)

Blenderiste09 committed:
Update copy.bat

Blenderiste09 committed:
Update copy.bat

Blenderiste09 committed:
Update copy.bat

Blenderiste09 committed:
Fix timing bug

Blenderiste09 committed:
Fix timing bug because of #$^*ù^$ demo format

Blenderiste09 committed:
Prepare the ghost managing

Changed name of Ghost.cpp/.hpp and added two files

Blenderiste09 committed:
Implemented Entity/Manager system

Blenderiste09 committed:
Ghost optimizations

Blenderiste09 committed:
Added different keyboard layout

sar_ihud_keyboard_layout

Blenderiste09 committed:
Fix crashes caused by deleting the ghost pointer twice

Blenderiste09 committed:
Changing model without restarting the level

Blenderiste09 committed:
Added trail to ghost

Only works with cheats on for now.
sar_ghost_trail_lenght
sar_ghost_trail_transparency

Blenderiste09 committed:
Demo record incrementing

sar_record_at
sar_record_at_demo_name
sar_record_at_increment

Blenderiste09 committed:
First import

Client can connect, disconnect, and stop the server.
Crashes when trying to update the ghost position

Blenderiste09 committed:
Cleaning code + fix

Blenderiste09 committed:
Removed Trail + some fixes

Demo record are dumbs. It breaks the trail

Blenderiste09 committed:
Merge branch 'master' of github.com:Blenderiste09/SourceAutoRecord

Blenderiste09 committed:
Fixing crashes + cleaning code

Blenderiste09 committed:
Fixed crashes and lag issues

Blenderiste09 committed:
Cleaning code

Blenderiste09 committed:
Added timeout

Blenderiste09 committed:
Cleanup

Blenderiste09 committed:
TCP socket for connection and disconnection

Blenderiste09 committed:
TCP fully working but SAR crash at sar_exit

Blenderiste09 committed:
Fix crash whe sar_exit

I'm big dumb

Blenderiste09 committed:
Cleanup

Blenderiste09 committed:
Ghost Server files

Blenderiste09 committed:
Name Fix

Blenderiste09 committed:
Fix error

Blenderiste09 committed:
Network optimization

Blenderiste09 committed:
Handle map change

Blenderiste09 committed:
Fix crashes because of optimization

Blenderiste09 committed:
ID fix

Blenderiste09 committed:
Implemented message sending. Need to fix it

Blenderiste09 committed:
Send model name to other

Blenderiste09 committed:
Fix some crashes

Blenderiste09 committed:
Remove local player ghost from ghostpool

Blenderiste09 committed:
Fix packet slow packet process

Blenderiste09 committed:
Pause the network if game is paused

Blenderiste09 committed:
Implemented linear interpolation

Allow player to run the game at lower tickrate without having the ghost jiterring

Blenderiste09 committed:
Don't send local player to local player

Blenderiste09 committed:
Automatic detection if ip address is local or public

Blenderiste09 committed:
Implemented synchronized countdown

Blenderiste09 committed:
Fix disconnection problems

Blenderiste09 committed:
Add model over the network

Blenderiste09 committed:
Fix ping and improved interpolation

Blenderiste09 committed:
Make CPU sleeping

Blenderiste09 committed:
Clean code

Blenderiste09 committed:
Display ghost names over them

Blenderiste09 committed:
Fix sar_record_at_demo_name

Blenderiste09 committed:
Clean up code

Blenderiste09 committed:
Fix wrong variable name

Blenderiste09 committed:
Clean up code

Blenderiste09 committed:
Start working on server GUI with TGUI (SFML lib)

Blenderiste09 committed:
Fixed sar_record_at_demo_name

Blenderiste09 committed:
Update on the GUI

Blenderiste09 committed:
New synchronized countdown system

Blenderiste09 committed:
Synchronized countdown and countdown + teleport

Blenderiste09 committed:
Clean code

Blenderiste09 committed:
Clean up

Blenderiste09 committed:
I'm dumb

Blenderiste09 committed:
Fix countdown function

Now deduce the ping. Can go up to 3000ms of ping

Blenderiste09 committed:
Fix sar_ghost_message crash

Blenderiste09 committed:
Fix crashes

Blenderiste09 committed:
Fix compiler warning

Blenderiste09 committed:
Clean up

Blenderiste09 committed:
Commands when requesting countdown

Blenderiste09 committed:
Stop lerp if time is > 1

Blenderiste09 committed:
ChatPrintf

Blenderiste09 committed:
Detect if level is loaded

Blenderiste09 committed:
Added pre/post countdown commands

Can be useful for synchronized teleport/load/pause

Blenderiste09 committed:
Clean up

Blenderiste09 committed:
Implemented sar_pause

Credits route to spt ;)

Blenderiste09 committed:
Clean up + Ghost demo player working again :)

Blenderiste09 committed:
Show names over network, demo, and demo playing ghosts

Blenderiste09 committed:
Clean up

Blenderiste09 committed:
Ghost not recorded in demos anymore

sar_ghost_type 1/2

Blenderiste09 committed:
Fixing invisible ghosts

Blenderiste09 committed:
Fix crash with playdemo in Portal 2

Blenderiste09 committed:
reduce loading times

ui_loadingscreen_fadein_time
ui_loadingscreen_mintransition_time

Credits route to Betsruner

Blenderiste09 committed:
Merge pull request #14 from Blenderiste09/ui_loadingscreen

reduce loading times

Blenderiste09 committed:
Merge branch 'master' into Network-ghosts

Blenderiste09 committed:
Revert "Ghost not recorded in demos anymore"

This reverts commit 4e8cf0c5bba364810678a1fc8146cf0f00dcd0a5.

Revert wrong manipulation

Blenderiste09 committed:
Merge branch 'Network-ghosts' of github.com:Blenderiste09/SourceAutoRecord into Network-ghosts

Blenderiste09 committed:
Revert to 7448146 (Fixing invisible ghosts)

Blenderiste09 committed:
Merge pull request #16 from Blenderiste09/Network-ghosts

Network ghosts

Blenderiste09 committed:
reduced loading times

ui_loadingscreen_fadein_time
ui_loadingscreen_mintransition_time

Credits route to Betsruner

Blenderiste09 committed:
Fix demo crash

Blenderiste09 committed:
Crosshair

NeKzor commented:
Closing this as we "discussed" that the ghost server code should not be part of SAR.
